arrow-core-data/src/main/kotlin/arrow/core/computations/suspended/NullableBindSyntax.kt

```diff
@@ -0,0 +1,12 @@
+package arrow.core.computations.suspended
+
+import arrow.continuations.generic.DelimitedScope
+
+data class NullableBindSyntax<R>(
+  private val scope: DelimitedScope<R?>,
+) : BindSyntax {
+  override suspend fun <A> A?.invoke(): A =
+    scope.shift { cont ->
+      this@invoke?.let { cont(it) }
+    }
```

Comment:
Can this be implemented to only call `shift` when `this` is `null` instead of calling it every time?

Reply:
🤔 `shift` promises to convert `A?` to `A` (nullable into a non-nullable). If my understanding is correct, scope.shift  is doing this with the continuation and if the value is null the continuation is not called.

Reply:
if the `invoke()` is executed on a `null` value there will be no more invoke function called. am I assuming this correctly?

just the illustrate with an example:
```
nullable {
val a = null.invoke() // last swift call, short circuit to the end

val b = a.invoke() // wont be called because of short circuit 
b.invoke() // wont be called because of short circuit 
} == null
```

Reply:
`shift` returns whatever is given to it's argument:
```kotlin
println(shift { it(3) }) // => 3
println(shift { it(9); it(5) }) // => 9\n5 (assuming multi shot is used)
```
So the return value of `shift` is what we pass to the continuation. But what if we never pass anything to it? Then it never returns anything: (The type of this is `Nothing` and infinite suspension, throwing and infinite recursion are the only ways to create a "value" of such a type. And since `Nothing` can be used in place of any type it does indeed also return `A`)
```kotlin
println(shift { _ -> null }) // => nothing here because we never resume
```
The function passed to `shift` has to return the final argument type of the delimited continuation `R` but how it does so doesn't actually matter. It is free to completely ignore the continuation passed to it.
```kotlin
val a: Int? = null
reset<Int?> {
  val b: Int = a ?: shift { null } // null is perfectly valid for `Int?` and `shift { null }` is `Nothing`
  b
} // => null
val a: Int? = null
reset<Int?> {
  val b: Int = a ?: shift { it(8) }
  b
} // => 8
```

There are a few things to note:
`shift { it(x) } == x` if, and only if, invoking the continuation is all you do (no other effects)
`shift { x }` has type `Nothing` since it short circuits
`shift { it(x); it(y) }` will first run the continuation with `x` and then with `y` (except we don't really support multishot so this will throw unless you use the multi shot continuation scope which, quite terribly, emulates this behavior)

Reply:
So the assumption that `shift` converts `A?` to `A` is wrong. It simply returns what is passed to the continuation. This is further evident by its type: `fun <A> shift(f: ((A) -> R) -> R): A`.

Reply:
ok,it started to click. Do you think we could include this explanation in the documentation? 

now I see the reason behind what you mentioned. 
> only call shift when this is null

Will apply the changes quickly

Reply:
Jannis, excellent disection of shift, agreed  we need something like that in the docs.